granularflow1pX.go

```diff
@@ -0,0 +1,375 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package edgedb
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/edgedb/edgedb-go/internal/buff"
+	"github.com/edgedb/edgedb-go/internal/cardinality"
+	"github.com/edgedb/edgedb-go/internal/codecs"
+	"github.com/edgedb/edgedb-go/internal/descriptor"
+	"github.com/edgedb/edgedb-go/internal/format"
+	"github.com/edgedb/edgedb-go/internal/message"
+)
+
+func (c *protocolConnection) execGranularFlow1pX(
+	r *buff.Reader,
+	q *query,
+) error {
+	ids, ok := c.getCachedTypeIDs(q)
+	if !ok {
+		return c.pesimistic1pX(r, q)
+	}
+
+	cdcs, err := c.codecsFromIDs(ids, q)
+	if err != nil {
+		return err
+	} else if cdcs == nil {
+		return c.pesimistic1pX(r, q)
+	}
+
+	return c.optimistic1pX(r, q, cdcs)
+}
+
+func (c *protocolConnection) pesimistic1pX(r *buff.Reader, q *query) error {
+	desc, err := c.parse1pX(r, q)
+	if err != nil {
+		return err
+	}
+
+	cdcs, err := c.codecsFromDescriptors1pX(q, desc)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.execute1pX(r, q, cdcs)
+	return err
+}
+
+func (c *protocolConnection) optimistic1pX(
+	r *buff.Reader,
+	q *query,
+	cdcs *codecPair,
+) error {
+	// When descriptors are returned the codec ids sent didn't match the
+	// server's.  The codecs should be rebuilt with the new descriptors and the
+	// execution retried.
```

Comment:
This is not correct.  When the output type descriptor is not matched, the client is supposed to receive the correct one and proceed decoding the data messages, no execution retry is necessary.

Reply:
Ok. What about if the input descriptor is wrong? Does that warrant an execution retry?

Reply:
For now, no, because it's a much rarer condition and is likely non-recoverable anyway.